File: nearbyfundi_backend/app/Models/Subscription.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\SoftDeletes;
use Carbon\Carbon;

class Subscription extends Model
{
    use SoftDeletes;

    protected $table = 'subscriptions';

    protected $fillable = [
        'user_id', 'rate_card_id', 'status', 'start_date', 'expiry_date',
        'amount_paid', 'currency', 'payment_method', 'payment_reference',
        'payment_proof', 'approved_at', 'approved_by', 'admin_notes'
    ];

    protected $casts = [
        'start_date' => 'datetime',
        'expiry_date' => 'datetime',
        'approved_at' => 'datetime',
        'amount_paid' => 'decimal:2',
    ];

    // Status constants
    const STATUS_PENDING = 'pending';
    const STATUS_APPROVED = 'approved';
    const STATUS_ACTIVE = 'active';
    const STATUS_EXPIRED = 'expired';
    const STATUS_CANCELLED = 'cancelled';

    // ============================================================
    // RELATIONSHIPS
    // ============================================================

    public function user()
    {
        return $this->belongsTo(User::class);
    }

    public function rateCard()
    {
        return $this->belongsTo(RateCard::class);
    }

    public function approver()
    {
        return $this->belongsTo(User::class, 'approved_by');
    }

    public function invoice()
    {
        return $this->hasOne(Invoice::class);
    }

    public function paymentMethod()
    {
        return $this->belongsTo(PaymentMethod::class, 'payment_method', 'name');
    }

    // ============================================================
    // STATUS HELPERS - FIXED
    // ============================================================

    /**
     * ✅ FIXED: Check if subscription is active and not expired
     */
    public function isActive(): bool
    {
        if ($this->status !== self::STATUS_ACTIVE) {
            return false;
        }
        
        // ✅ If expiry_date is null, consider it active
        if ($this->expiry_date === null) {
            return true;
        }
        
        // ✅ Ensure expiry_date is a Carbon instance
        $expiry = $this->expiry_date;
        if ($expiry instanceof \DateTime) {
            return $expiry->isFuture();
        }
        
        return false;
    }

    /**
     * ✅ FIXED: Check if subscription is expired
     */
    public function isExpired(): bool
    {
        if ($this->status === self::STATUS_EXPIRED) {
            return true;
        }
        
        if ($this->expiry_date === null) {
            return false;
        }
        
        $expiry = $this->expiry_date;
        if ($expiry instanceof \DateTime) {
            return $expiry->isPast();
        }
        
        return false;
    }

    public function isPending(): bool
    {
        return $this->status === self::STATUS_PENDING;
    }

    public function isApproved(): bool
    {
        return $this->status === self::STATUS_APPROVED;
    }

    public function isCancelled(): bool
    {
        return $this->status === self::STATUS_CANCELLED;
    }

    // ============================================================
    // ACCESSORS
    // ============================================================

    /**
     * ✅ FIXED: Get days remaining safely
     */
    public function getDaysRemainingAttribute(): ?int
    {
        if (!$this->expiry_date) {
            return null;
        }
        
        $expiry = $this->expiry_date;
        if ($expiry instanceof \DateTime) {
            if ($expiry->isPast()) {
                return 0;
            }
            return $expiry->diffInDays(now());
        }
        
        return null;
    }

    public function getStatusLabelAttribute(): string
    {
        return match($this->status) {
            self::STATUS_PENDING => 'Pending Approval',
            self::STATUS_APPROVED => 'Approved',
            self::STATUS_ACTIVE => 'Active',
            self::STATUS_EXPIRED => 'Expired',
            self::STATUS_CANCELLED => 'Cancelled',
            default => ucfirst($this->status),
        };
    }

    // ============================================================
    // SCOPES
    // ============================================================

    public function scopeActive($query)
    {
        return $query->where('status', self::STATUS_ACTIVE)
                     ->where(function($q) {
                         $q->whereNull('expiry_date')
                           ->orWhere('expiry_date', '>', now());
                     });
    }

    public function scopeExpired($query)
    {
        return $query->where('status', self::STATUS_EXPIRED)
                     ->orWhere(function($q) {
                         $q->where('status', self::STATUS_ACTIVE)
                           ->where('expiry_date', '<=', now());
                     });
    }

    public function scopePending($query)
    {
        return $query->where('status', self::STATUS_PENDING);
    }

    public function scopeApproved($query)
    {
        return $query->where('status', self::STATUS_APPROVED);
    }

    public function scopeCancelled($query)
    {
        return $query->where('status', self::STATUS_CANCELLED);
    }

    // ============================================================
    // EVENTS
    // ============================================================

    protected static function booted()
    {
        static::saving(function ($subscription) {
            if ($subscription->expiry_date && $subscription->expiry_date->isPast()) {
                $subscription->status = self::STATUS_EXPIRED;
            }
        });
    }
}

File: fundiapp/lib/models/subscription.dart
import 'rate_card.dart';
import 'invoice.dart';

class Subscription {
  final int id;
  final int userId;
  final RateCard? rateCard;
  final String status;
  final String statusLabel;
  final DateTime? startDate;
  final DateTime? expiryDate;
  final int? daysRemaining;
  final String amountPaid;
  final String currency;
  final String? paymentMethod;
  final String? paymentReference;
  final String? paymentProof;
  final DateTime? approvedAt;
  final String? approvedBy;
  final String? adminNotes;
  final DateTime createdAt;
  final Invoice? invoice;

  Subscription({
    required this.id,
    required this.userId,
    this.rateCard,
    required this.status,
    required this.statusLabel,
    this.startDate,
    this.expiryDate,
    this.daysRemaining,
    required this.amountPaid,
    required this.currency,
    this.paymentMethod,
    this.paymentReference,
    this.paymentProof,
    this.approvedAt,
    this.approvedBy,
    this.adminNotes,
    required this.createdAt,
    this.invoice,
  });

  factory Subscription.fromJson(Map<String, dynamic> json) {
    return Subscription(
      id: json['id'] ?? 0,
      userId: json['user_id'] ?? 0,
      rateCard: json['rate_card'] != null ? RateCard.fromJson(json['rate_card']) : null,
      status: json['status'] ?? 'pending',
      statusLabel: json['status_label'] ?? 'Pending',
      startDate: json['start_date'] != null ? DateTime.parse(json['start_date']) : null,
      expiryDate: json['expiry_date'] != null ? DateTime.parse(json['expiry_date']) : null,
      daysRemaining: json['days_remaining'],
      amountPaid: json['amount_paid']?.toString() ?? '0',
      currency: json['currency'] ?? 'TZS',
      paymentMethod: json['payment_method'],
      paymentReference: json['payment_reference'],
      paymentProof: json['payment_proof'],
      approvedAt: json['approved_at'] != null ? DateTime.parse(json['approved_at']) : null,
      approvedBy: json['approved_by'],
      adminNotes: json['admin_notes'],
      createdAt: json['created_at'] != null ? DateTime.parse(json['created_at']) : DateTime.now(),
      invoice: json['invoice'] != null ? Invoice.fromJson(json['invoice']) : null,
    );
  }

  // ✅ FIXED: Check both status AND expiry date
  bool get isPending => status == 'pending';

  bool get isActive => status == 'active' &&
      (expiryDate == null || expiryDate!.isAfter(DateTime.now()));

  bool get isExpired => status == 'expired' ||
      (status == 'active' && expiryDate != null && expiryDate!.isBefore(DateTime.now()));

  bool get isCancelled => status == 'cancelled';

  bool get isApproved => status == 'approved';

  String get statusMessage {
    if (isActive) return 'Active';
    if (isPending) return 'Pending Approval';
    if (isExpired) return 'Expired';
    if (isCancelled) return 'Cancelled';
    return statusLabel;
  }
}